# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/datdigitaladdict/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

I'm picking Lane 2: Refresh / Content Opportunity Scoring. I've already run the starter pipeline in notebooks 01 and 02 and seen a learned model meaningfully beat a hand-written rule at ranking pages for review, which makes me want to dig deeper into how a scoring system like this could actually guide someone's daily work.

In [1]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

Unit of analysis: one content page (content_id).
Decision: which pages should a content reviewer look at first, out of potentially thousands, when they only have time to review a limited number per week.
Action: a human reviewer opens the top-ranked pages and decides whether to refresh, expand, or leave them, based on reason codes like stale_visible_page or declining_with_demand.
Cost of a wrong call: if a page is wrongly flagged as high-priority, the reviewer wastes limited time on a page that didn't need attention, meanwhile a page that's actually declining gets missed. Write these as separate labeled points or short paragraphs, not as one run-on paragraph.

In [2]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

Here are three numbers from the starter dataset supporting this lane:

In [3]:
import os, sys, subprocess

IN_COLAB = "google.colab" in sys.modules
REPO_URL = "https://github.com/datdigitaladdict/flyrank-ml-internship"
REPO_DIR = "flyrank-ml-internship"

if IN_COLAB:
    if not os.path.isdir(REPO_DIR):
        subprocess.run(["git", "clone", "--depth", "1", REPO_URL, REPO_DIR], check=True)
    os.chdir(REPO_DIR)

print("Working dir:", os.getcwd())
assert os.path.exists("data/raw/content_refresh_anonymized.csv"), "starter CSV not found"
print("Starter data found.")

Working dir: /content/flyrank-ml-internship
Starter data found.


Three numbers from the starter dataset show why this lane is worth pursuing: how much of the inventory is currently declining, how narrow a simple rule-based filter is, and how much a trained model can improve targeting over a hand-written rule.

In [4]:
import pandas as pd, json

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")
df["is_declining_label"] = df["trend_direction"].str.lower().eq("down").astype(int)

declining_rate = df["is_declining_label"].mean()
print(f"Share of pages currently labeled 'declining': {declining_rate:.3f}")

stale_visible = df[(df["days_since_last_update"] >= 180) & (df["impressions_90d"] >= 500)]
print(f"Pages meeting 'stale + visible' criteria: {len(stale_visible)} of {len(df)}")

res = json.load(open("outputs/model_results.json"))
base = res["baseline"]["baseline_precision_at_50"]
rf = res["models"]["random_forest"]["precision_at_50"]
print(f"Hand-rule Precision@50: {base:.3f}   Random forest Precision@50: {rf:.3f}")

Share of pages currently labeled 'declining': 0.542
Pages meeting 'stale + visible' criteria: 17 of 30000
Hand-rule Precision@50: 0.240   Random forest Precision@50: 0.740


## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

These numbers show real patterns, not proof of cause. 54.2% of pages are labeled "declining," but only 17 of 30,000 pages meet a simple stale-and-visible rule, showing that a fixed rule alone is far too narrow to be useful at scale. The random forest reached 0.740 Precision@50 versus the hand rule's 0.240, showing a learned ranking can meaningfully outperform a fixed rule on this starter slice. This does not mean refreshing a page will cause it to recover, that would require an actual before/after experiment. It also doesn't rule out other explanations for decline, like consolidation or seasonality, which I have not checked yet. These results come from a small 30,000-row anonymized sample using client-holdout validation, not the full ~79M-row warehouse, so they should be read as directional and decision-support only, not as a finished, generalizable model.

In [5]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.